src/main/java/com/streamr/client/dataunion/DataUnion.java

```diff
@@ -0,0 +1,233 @@
+    private static final Logger log = LoggerFactory.getLogger(DataUnion.class);
+    public enum ActiveStatus{none, active, inactive};
+
+    private DataUnionMainnet mainnet;
+    private DataUnionSidechain sidechain;
+    private Web3j mainnetConnector, sidechainConnector;
+    private Credentials mainnetCred, sidechainCred;
+
+    //use DataUnionClient to instantiate
+    protected DataUnion(DataUnionMainnet mainnet, Web3j mainnetConnector, Credentials mainnetCred, DataUnionSidechain sidechain, Web3j sidechainConnector, Credentials sidechainCred) {
+        this.mainnet = mainnet;
+        this.mainnetConnector = mainnetConnector;
+        this.mainnetCred = mainnetCred;
+        this.sidechain = sidechain;
+        this.sidechainConnector = sidechainConnector;
+        this.sidechainCred = sidechainCred;
+    }
+
+    public boolean waitForDeployment(long pollInterval, long timeout) throws Exception {
+        String code = waitForCodeAtAddress(sidechain.getContractAddress(), sidechainConnector, pollInterval, timeout);
+        return code != null && !code.equals("0x");
+    }
+
+    public String mainnetAddress(){
+        return mainnet.getContractAddress();
+    }
+
+    public String sidechainAddress(){
+        return sidechain.getContractAddress();
+    }
+
+    public boolean isDeployed() throws Exception {
+        //this will check the condition only once:
+        return waitForDeployment(0,0);
+    }
+
+    public void sendTokensToBridge() throws Exception {
+        mainnet.sendTokensToBridge().send();
+    }
+
+    public BigInteger waitForEarningsChange(final BigInteger initialBalance, long pollInterval, long timeout) throws Exception {
+        Web3jUtils.Condition earningsChange = new Web3jUtils.Condition(){
+            @Override
+            public Object check() throws Exception {
+                BigInteger bal = sidechain.totalEarnings().send().getValue();
+                if(!bal.equals(initialBalance)) {
+                    return bal;
+                }
+                return null;
+            }
+        };
+        return (BigInteger) waitForCondition(earningsChange, pollInterval, timeout);
+    }
+
+    public EthereumTransactionReceipt addJoinPartAgents(String ... agents) throws Exception {
+        return new EthereumTransactionReceipt(sidechain.addJoinPartAgents(asDynamicAddressArray(agents)).send());
+    }
+
+    public EthereumTransactionReceipt partMembers(String ... members) throws Exception {
+        return new EthereumTransactionReceipt(sidechain.partMembers(asDynamicAddressArray(members)).send());
+    }
+
+    public EthereumTransactionReceipt joinMembers(String ... members) throws Exception {
+        return new EthereumTransactionReceipt(sidechain.addMembers(asDynamicAddressArray(members)).send());
+    }
+
+
+    /**
+     *
+     * withdraw a member to his own address. Must be that member or admin.
+
+     * @param member
+     * @param amount amout in wei or 0 to withdraw everything
```

Comment:
If zero is a magic number, we could introduce a constant for that. Or as this is a very important parameter, we could even use a separate method for withdrawing everything. (E.g. if a user inputs 0 in an UI, he wouldn't expect anything to happen.)

Reply:
added withdrawAll.. methods

Reply:
I'd add an assertion to public `withdraw` methods so that they would accept only amounts that are greater than `0`.

Reply:
ok